File: cpp/hardware/storage/StorageDevice.cpp
#pragma once
#include <bits/stdc++.h>

#include "../../component/Component.cpp"
#include "../../component/Hardware.cpp"
#include "../../interfaces/Connectable.cpp"

class StorageDevice : public Hardware, public Connectable, public Component
{
protected:
    int capacityGB;
    int readSpeedMBs;
    int writeSpeedMBs;

public:
    // Constructor
    StorageDevice() : Hardware(), Component(), capacityGB(0), readSpeedMBs(0), writeSpeedMBs(0) {}
    StorageDevice(const std::string &serial, const Date &installDate, bool installed, int capacity, int readSpeed, int writeSpeed, const std::string &id, const std::string &manufacturer, const std::string &model, double price)
        : Hardware(serial, installDate, installed), Component(id, manufacturer, model, price), capacityGB(capacity), readSpeedMBs(readSpeed), writeSpeedMBs(writeSpeed) {}

    // Destructor
    ~StorageDevice() {}

    // Setters
    void setCapacityGB(int capacity)
    {
        capacityGB = capacity;
    }

    void setReadSpeedMBs(int readSpeed)
    {
        readSpeedMBs = readSpeed;
    }

    void setWriteSpeedMBs(int writeSpeed)
    {
        writeSpeedMBs = writeSpeed;
    }

    // Getters
    int getCapacityGB() const
    {
        return capacityGB;
    }

    int getReadSpeedMBs() const
    {
        return readSpeedMBs;
    }

    int getWriteSpeedMBs() const
    {
        return writeSpeedMBs;
    }

    bool connect() override
    {
        printf("Storage device connected\n");
        return true;
    }

    bool disconnect() override
    {
        printf("Storage device disconnected\n");
        return true;
    }
};

File: python/hardware/storage/storage_device.py
from datetime import date
from components.component import Component
from components.hardware import Hardware
from interfaces.connectable import Connectable

class StorageDevice(Hardware, Connectable, Component):
    def __init__(self, serial_number="", installation_date=None, physically_installed=False, capacity_gb=0, read_speed_mbs=0, write_speed_mbs=0, id="", manufacturer="", model="", price=0.0):
        Hardware.__init__(self, serial_number, installation_date, physically_installed)
        Component.__init__(self, id, manufacturer, model, price)
        self._capacity_gb = capacity_gb
        self._read_speed_mbs = read_speed_mbs
        self._write_speed_mbs = write_speed_mbs

    # Setters
    def set_capacity_gb(self, capacity_gb):
        self._capacity_gb = capacity_gb

    def set_read_speed_mbs(self, read_speed_mbs):
        self._read_speed_mbs = read_speed_mbs

    def set_write_speed_mbs(self, write_speed_mbs):
        self._write_speed_mbs = write_speed_mbs

    # Getters
    def get_capacity_gb(self):
        return self._capacity_gb

    def get_read_speed_mbs(self):
        return self._read_speed_mbs

    def get_write_speed_mbs(self):
        return self._write_speed_mbs

    def connect(self):
        print("Storage device connected")
        return True

    def disconnect(self):
        print("Storage device disconnected")
        return True